src/metabase/query_processor/middleware/optimize_temporal_filters.clj

```diff
@@ -104,75 +115,253 @@
+
+(defmethod optimize :between
+  [[_between x lower upper]]
+  (let [x     (optimize x)
+        lower (optimize lower)
+        upper (optimize upper)]
+    (or (case [(temporal-category x) (temporal-category lower) (temporal-category upper)]
+          [::ref ::literal ::literal]
+          (let [x-unit     (non-default-temporal-unit x)
+                lower-unit (non-default-temporal-unit lower)
+                upper-unit (non-default-temporal-unit upper)
+                unit       (or x-unit lower-unit upper-unit)]
+            (when (and unit
+                       (optimizable-units unit)
+                       (same-units-or-nil? [x-unit lower-unit upper-unit]))
+              (let [x' (change-temporal-unit-to-default x)]
+                [:and
+                 [:>= x' (temporal-value-lower-bound lower unit)]
+                 [:<  x' (temporal-value-upper-bound upper unit)]])))
+
+          [::arithmetic-wrapping-ref ::literal ::literal]
+          (let [{x' :x, lower' :y} (transfer-temporal-arithmetic-args x lower)
+                {upper' :y}        (transfer-temporal-arithmetic-args x upper)]
+            (optimize [:between x' lower' upper']))
+          #_else
+          nil)
+        [:between x lower upper])))
+
+(mu/defn ^:private combine-intervals-if-possible :- [:maybe mbql.s/interval]
+  "Combine two `:interval` clauses into one if they both have the same unit."
+  ([[_interval x-n x-unit] :- mbql.s/interval
+    [_interval y-n y-unit] :- mbql.s/interval]
+   (when (= x-unit y-unit)
+     [:interval (+ x-n y-n) x-unit])))
+
+(mu/defn optimize-intervals :- [:sequential {:min 1} mbql.s/interval]
+  "Combine multiple `:interval` clauses into the fewest number possible."
+  [intervals :- [:sequential {:min 1} mbql.s/interval]]
+  (let [intervals (sort-by
+                   (fn [[_interval _n unit]]
+                     unit)
+                   intervals)]
+    (reduce
+     (fn [acc interval]
+       (if (empty? acc)
+         [interval]
+         (if-let [combined (combine-intervals-if-possible (last acc) interval)]
+           (conj (vec (butlast acc)) combined)
+           (conj (vec acc) interval))))
+     []
+     intervals)))
+
+(mu/defn ^:private add-interval-if-possible :- [:maybe [:or mbql.s/relative-datetime mbql.s/absolute-datetime]]
+  "Combine an `:absolute-datetime`/`:relative-datetime` and an `:interval` clause, if possible."
+  [literal                              :- [:or mbql.s/relative-datetime mbql.s/absolute-datetime]
+   [_interval interval-n interval-unit] :- mbql.s/interval]
+  (case (first literal)
+    :absolute-datetime
+    (let [[_absolute-datetime t unit] literal]
+      (when (= unit interval-unit)
+        [:absolute-datetime (u.date/add t unit interval-n) unit]))
+
+    :relative-datetime
+    (let [[_relative-datetime n unit] literal]
+      (if (= n :current)
+        [:relative-datetime interval-n interval-unit]
```

Comment:
I can't convince myself that this is sound in all cases, but maybe I'm thinking of it wrong.

My vague mental model is that a `:relative-datetime` defines a window of time (say a month) and the interval gives a delta like `+ 3 weeks`. In my mind adding 3 weeks to "this month" turns `Nov 1 - Nov 30, 2023` into `Nov 22 - Dec 21, 2023`.

But if I'm understanding how `:relative-datetime` works properly, this code is dropping the original window size (month) and replacing it with "3 weeks from this week". So for example, starting from Friday Nov 24 and using Monday as the first day, that rounds up to the next week (Mon Nov 27) and then adds two more, resulting in Mon Dec 11 to Sun Dec 17.

Do we want it to translate a same-sized window, or is changing the window intended? Or is there a missing `(= unit interval-unit)` check for this case?

Reply:
ended up removing this stuff